src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3502,86 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    /// Looks up a candidate entities with a given name and filters them using
+    /// the supplied predicate. The lookup spans across all crates. This method
+    /// should be used to generate suggestions when name resolution fails.
+    fn lookup_candidates<FilterFn>(&mut self,
+                                   lookup_name: Name,
+                                   namespace: Namespace,
+                                   filter_fn: FilterFn) -> SuggestedCandidates
+        where FilterFn: Fn(Def) -> bool {
+
+        let mut lookup_results = Vec::new();
+        let mut lookup_modules = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new()));
+
+        while let Some((in_module, path_segments)) = lookup_modules.pop() {
+            build_reduced_graph::populate_module_if_necessary(self, &in_module);
+
+            in_module.for_each_child(|name, ns, name_binding| {
+
+                // avoid imports entirely
+                if !name_binding.is_import() {
+
+                    // collect results based on the filter function
+                    if let Some(def) = name_binding.def() {
+                        if ns == namespace {
+                            if filter_fn(def) && name == lookup_name {
```

Comment:
Ok, last nit: replace the above two lines with `if name == lookup_name && ns == namespace && filter_fn(def)` to reduce rightward drift.